cli/flox/src/utils/openers.rs

```diff
@@ -71,6 +74,92 @@ impl Browser {
     }
 }
 
+#[derive(Debug)]
+pub enum Shell {
+    Bash(PathBuf),
+    Zsh(PathBuf),
+}
+
+impl TryFrom<&Path> for Shell {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &Path) -> std::prelude::v1::Result<Self, Self::Error> {
+        match value.file_name() {
+            Some(name) if name == "bash" => Ok(Shell::Bash(value.to_owned())),
+            Some(name) if name == "zsh" => Ok(Shell::Zsh(value.to_owned())),
+            _ => Err(anyhow!("Unsupported shell {value:?}")),
+        }
+    }
+}
+
+impl Display for Shell {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Shell::Bash(_) => write!(f, "bash"),
+            Shell::Zsh(_) => write!(f, "zsh"),
+        }
+    }
+}
+
+impl Shell {
+    /// Detect the current shell from the parent process
+    ///
+    /// This function tries to detect the shell from the parent process.
+    /// If reading process information of the parent process fails,
+    /// or the exe path of the parent process can not be parsed to a known shell,
+    /// an error is returned.
+    fn detect_from_parent_process() -> Result<Self> {
+        // todo: we can narrow down the amount of data collected by sysinfo, for now collect everything
+        let system = System::new_all();
+
+        let parent_process = system
+            .process(Pid::from_u32(std::os::unix::process::parent_id()))
+            .context("Failed to get info about parent process")?;
+
+        let parent_exe = parent_process
+            .exe()
+            .context("Failed to get parent process exe")?;
+
+        let shell = Self::try_from(parent_exe)?;
+        Ok(shell)
+    }
+
+    /// Detect the current shell from the {var} environment variable
+    fn detect_from_env(var: &str) -> Result<Self> {
+        let shell = env::var(var)
+            .with_context(|| format!("{var} environment variable not set"))
+            .and_then(|shell| {
+                let path = PathBuf::from(shell);
+                Self::try_from(path.as_path())
+            })?;
+        Ok(shell)
```

Comment:
nit - unless maybe you think `Ok(shell)` is more readable?
```suggestion
        env::var(var)
            .with_context(|| format!("{var} environment variable not set"))
            .and_then(|shell| {
                let path = PathBuf::from(shell);
                Self::try_from(path.as_path())
            })
```